src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -75,16 +79,26 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
 
   override def settings = currentSettings
 
-  private[this] var currentReporter: Reporter = { reporter = reporter0 ; currentReporter }
+  // Umad reported violation on: `scala.reflect.internal.SymbolTable.scala$reflect$internal$Names$$nc_$eq(int)`
+  // To synchronize access we can re-use mechanism from `scala.reflect.internal.Names` so far used for runtime mirror.
+  // For efficiency want to enable it only for phases which are parallelized.
+  override protected def synchronizeNames: Boolean = _synchronizeNames
+  private[this] var _synchronizeNames = false
 
-  def reporter: Reporter = currentReporter
+  /* `currentReporter` is used on both main thread as well as worker threads.
+   * On worker threads for every unit we are creating new `BufferedReporter`
+   * which at the end of the unit processing is dumped to the main reporter.
+   * We need to do it if we want to retain the same messages order as in case of single threaded execution.
+   */
+  private[this] val currentReporter: WorkerOrMainThreadLocal[Reporter] = WorkerThreadLocal(reporter0, reporter0)
+  def reporter: Reporter =  currentReporter.get
   def reporter_=(newReporter: Reporter): Unit =
-    currentReporter = newReporter match {
+    currentReporter.set(newReporter match {
```

Comment:
we should assert that we are on the main thread I think

Reply:
That is not the case here, please look into `processUnit` method.